### PR TITLE
Martin/jib 133 display avatar with line

### PR DESCRIPTION
**Pull request by MartinJibber**
There are spaces between the lines and they're not colored in yet, but it's a start.

Simulator Screen Shot - iPhone 12 - 2021-10-19 at 16 02 16

### CODE REVIEW

File: iOS/Flows/Conversation/Cells/MessageCell.swift [hunk 1]
```diff
@@ -54,7 +71,41 @@ class MessageCell: UICollectionViewCell {
...
+        self.collectionView.collectionViewLayout.invalidateLayout()
+        if shouldShowAvatar {
+            self.collectionView.left = 40
+        } else {
+            self.collectionView.left = 0
+        }
+        self.collectionView.expand(.right)
+        self.collectionView.expandToSuperviewHeight()
+
+        if shouldShowAvatar {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessageCell.swift:87]: Different layouts depending on whether the avatar is shown.

File: iOS/Flows/Conversation/Cells/MessageCell.swift [hunk 2]
```diff
@@ -70,12 +121,25 @@ class MessageCell: UICollectionViewCell {
...
+    func setAuthor(with avatar: Avatar, showTopLine: Bool, showBottomLine: Bool) {
+        self.authorView.set(avatar: avatar)
+        self.authorView.isVisible = showBottomLine && !showTopLine
+
+        self.topVerticalLine.isHidden = !showTopLine
+        self.bottomVerticalLine.isHidden = !showBottomLine
+
+        self.dotView.isVisible = showTopLine && !showBottomLine
+
+        self.setNeedsLayout()
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessageCell.swift:134]: Logic for determining which lines to show.

File: iOS/Flows/Conversation/ConversationViewControllerDataSource+CellRegistration.swift [hunk 3]
```diff
@@ -39,6 +39,41 @@ extension ConversationCollectionViewDataSource {
         }
     }
 
+    static func createThreadMessageCellRegistration() -> MessageCellRegistration {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationViewControllerDataSource+CellRegistration.swift:42]: There's a different cell registration for the threaded style.

File: iOS/Flows/Conversation/ConversationViewControllerDataSource.swift [hunk 4]
```diff
@@ -50,12 +59,20 @@ class ConversationCollectionViewDataSource: CollectionViewDataSource<Conversatio
...
 
-            let cell = collectionView.dequeueConfiguredReusableCell(using: self.messageCellRegistration,
+            let cell: MessageCell
+            switch self.messageStyle {
+            case .conversation:
+                cell = collectionView.dequeueConfiguredReusableCell(using: self.messageCellRegistration,
+                                                                    for: indexPath,
+                                                                    item: (channelID, messageID, self))
+            case .thread:
+                cell = collectionView.dequeueConfiguredReusableCell(using: self.threadMessageCellRegistration,
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationViewControllerDataSource.swift]: In the data source we switch on the style of conversation.

File: iOS/Flows/Conversation/ConversationViewControllerDataSource.swift [hunk 5]
```diff
@@ -50,9 +59,17 @@ class ConversationCollectionViewDataSource: CollectionViewDataSource<Conversatio
...
+            let cell: MessageCell
+            switch self.messageStyle {
+            case .conversation:
+                cell = collectionView.dequeueConfiguredReusableCell(using: self.messageCellRegistration,
                                                                     for: indexPath,
                                                                     item: (channelID, messageID, self))
+            case .thread:
+                cell = collectionView.dequeueConfiguredReusableCell(using: self.threadMessageCellRegistration,
+                                                                    for: indexPath,
+                                                                    item: (channelID, messageID, self))
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/ConversationViewControllerDataSource.swift:71]: Switching on the conversation type to determine which cell registration to use.

File: iOS/_Stream Temp/ChatMessageController+Extensions.swift [hunk 6]
```diff
@@ -13,6 +13,14 @@ typealias MessageController = ChatMessageController
...
 extension ChatMessageController {
 
+    var parentMessage: Message? {
+        guard let parentMessageID = self.message?.parentMessageId else { return nil }
+        let parentMessageController = ChatClient.shared.messageController(cid: self.cid,
+                                                                          messageId: parentMessageID)
+
+        return parentMessageController.message
+    }
+
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ChatMessageController+Extensions.swift:23]: Convenient way to get the parent message for replies.

File: iOS/Flows/Conversation/Cells/MessageCell.swift [hunk 7]
```diff
@@ -23,6 +23,11 @@ class MessageCell: UICollectionViewCell {
         cell.setText(with: item)
     }
 
+    private let authorView = AvatarView()
+    private let topVerticalLine = UIView()
```
Comment (Benjamindodgson) [inline iOS/Flows/Conversation/Cells/MessageCell.swift]: Dont use UIView

**Review by Benjamindodgson — APPROVED** [review]
Just change those to Views

**Pull request merged**